modin/core/dataframe/base/exchange/dataframe_protocol/from_dataframe.py

```diff
@@ -0,0 +1,463 @@
+# Licensed to Modin Development Team under one or more contributor license agreements.
+# See the NOTICE file distributed with this work for additional information regarding
+# copyright ownership.  The Modin Development Team licenses this file to you under the
+# Apache License, Version 2.0 (the "License"); you may not use this file except in
+# compliance with the License.  You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software distributed under
+# the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF
+# ANY KIND, either express or implied. See the License for the specific language
+# governing permissions and limitations under the License.
+
+"""Module houses functions building a ``pandas.DataFrame`` from DataFrame exchange protocol object."""
+
+import pandas
+import ctypes
+import numpy as np
+
+from typing import Optional, Tuple, Any, Union
+from modin.core.dataframe.base.exchange.dataframe_protocol.utils import (
+    DTypeKind,
+    ColumnNullType,
+)
+from modin.core.dataframe.base.exchange.dataframe_protocol.dataframe import (
+    ProtocolDataframe,
+    ProtocolColumn,
+    ProtocolBuffer,
+)
+
+
+def from_dataframe(
+    df: ProtocolDataframe, allow_copy: bool = True, nchunks: Optional[int] = None
+):
+    """
+    Build ``pandas.DataFrame`` from an object supporting DataFrame exchange protocol (__dataframe__).
```

Comment:
Modin core dataframe is built via `QueryCompiler.from_dataframe` method that uses this `from_dataframe` function in its basic implementation. In the future, each `QueryCompiler.from_dataframe` should implement its own import function with zero-copy when possible.

Reply:
Look at this discussion ([link](https://github.com/modin-project/modin/pull/4269#discussion_r826786563)).

Reply:
I am still for building a Modin Dataframe, which is specific to the execution. `from_dataframe` should build a dataframe specific to the library that implements it. We discussed [this](https://github.com/data-apis/dataframe-api/issues/42#:~:text=We%20recommend%20(but%20don%27t%20require)%20adding%20a%20standalone%20factory%20function%20for%20this%20purpose.%20IF%20an%20implementer%20decides%20to%20do%20this%2C%20please%20use%20the%20from_dataframe%20signature%20given%20above.%20And%20if%20you%20need%20extensions%2C%20come%20talk%20to%20us.) with you.

Reply:
How do you expect this function to be used by `modin_frame.from_dataframe` then? Most of the implementations would like to use from-dataframe-to-pandas conversion in their initial implementation of `modin_frame.from_dataframe`. It will be impossible to do so if this `from_dataframe` utility function be building a Modin DataFrame depending on the selected execution.

We already have `from_dataframe` implementation that **"builds a dataframe specific to the library that implements it"**, it's `modin.pandas.utils.from_dataframe` - the function that builds a Modin DataFrame (a dataframe specific to the library (modin) that implements it), here we have `modin.core.dataframe.base.exchange.dataframe_protocol.from_dataframe` internal utility function, that's **for internal usage only**. 

If I rename this instance `from_dataframe -> from_dataframe_to_pandas` would that be okay?

Reply:
> How do you expect this function to be used by `modin_frame.from_dataframe` then? Most of the implementations would like to use from-dataframe-to-pandas conversion in their initial implementation of `modin_frame.from_dataframe`. It will be impossible to do so if this `from_dataframe` utility function be building a Modin DataFrame depending on the selected execution.

Just import a standalone `from_dataframe` function inside of `modin_frame.from_dataframe` and return the result from it. We can build a pandas DataFrame inside of `from_dataframe` and use `from_pandas` to construct a Modin Dataframe for `PandasOn*` execution. Similar logic is true for `OmnisciOnNative` execution, where we likely want to build a pyarrow table inside of `from_dataframe` and use `from_arrow` to construct a Modin Dataframe.

> We already have `from_dataframe` implementation that **"builds a dataframe specific to the library that implements it"**, it's `modin.pandas.utils.from_dataframe` - the function that builds a Modin DataFrame (a dataframe specific to the library (modin) that implements it), here we have `modin.core.dataframe.base.exchange.dataframe_protocol.from_dataframe` internal utility function, that's **for internal usage only**.

Modin core dataframe layer can be used as a standalone solution. We implement the protocol exactly for this layer.

Reply:
In offline discussion agreed to leave things as it is. 

Moved `from_dataframe.py` from `dataframe/base/exchange` to `dataframe/pandas/exchange`.